File: domains/ai/data-forge/tools/llm.ts
/**
 * DataForge LLM Tool — Cloud LLM text generation and image description.
 *
 * Wraps ``python -m data_forge.llm`` CLI, which uses any OpenAI-compatible
 * API provider (OpenAI, DeepSeek, Groq, Together, vLLM, Ollama, etc.).
 *
 * Subcommands:
 *   chat            — Single-turn text generation
 *   describe        — Image description via vision model
 *   batch-chat      — Batch text generation from prompt files
 *   batch-describe  — Batch image description
 *   list-providers  — List configured providers from keyfile
 */
import { tool } from "@opencode-ai/plugin"
import path from "path"

// ---------------------------------------------------------------------------
// Tool definition
// ---------------------------------------------------------------------------

const data_forge_llm = tool({
  description:
    "Cloud LLM text generation and image description via OpenAI-compatible APIs. " +
    "Supports chat, describe, batch-chat, batch-describe, and list-providers subcommands. " +
    "Requires a JSON keyfile with provider credentials (or KEYFILE environment variable).",

  args: {
    // ── Subcommand ────────────────────────────────────────────────────
    subcommand: tool.schema
      .enum(["chat", "describe", "batch-chat", "batch-describe", "list-providers"])
      .describe(
        "Operation to perform: chat = single text generation; describe = image description; " +
          "batch-chat = batch text generation from .txt files; " +
          "batch-describe = batch image description; list-providers = show available providers.",
      ),

    // ── Core (required per subcommand) ─────────────────────────────────
    provider: tool.schema
      .string()
      .optional()
      .describe(
        "Provider name defined in keyfile, e.g. 'openai', 'deepseek'. Required for chat, describe, batch-chat, batch-describe.",
      ),
    model: tool.schema
      .string()
      .optional()
      .describe(
        "Model name, e.g. 'gpt-4o', 'deepseek-chat'. Required for chat, describe, batch-chat, batch-describe.",
      ),

    // ── Chat / describe ────────────────────────────────────────────────
    prompt: tool.schema
      .string()
      .optional()
      .describe(
        "User prompt text. Required for 'chat'. Optional for 'describe' and 'batch-describe' (defaults to image description prompt).",
      ),
    image: tool.schema
      .string()
      .optional()
      .describe("Path to an image file. Required for 'describe' subcommand."),

    // ── Batch operations ───────────────────────────────────────────────
    inputDir: tool.schema
      .string()
      .optional()
      .describe(
        "Input directory: .txt prompt files for 'batch-chat', image files for 'batch-describe'.",
      ),
    outputDir: tool.schema
      .string()
      .optional()
      .describe("Output directory for batch results."),

    // ── Keyfile ────────────────────────────────────────────────────────
    keyfile: tool.schema
      .string()
      .optional()
      .describe(
        "Path to a JSON keyfile mapping provider names to {api_key, base_url}. " +
          "If omitted, the KEYFILE environment variable is used. Required for all subcommands.",
      ),

    // ── Optional tunables ──────────────────────────────────────────────
    system: tool.schema
      .string()
      .optional()
      .describe("System message. Supported by 'chat' and 'batch-chat' subcommands."),
    temperature: tool.schema
      .number()
      .optional()
      .default(0.7)
      .describe("Sampling temperature (0.0–2.0, default: 0.7)."),
    maxTokens: tool.schema
      .number()
      .optional()
      .describe("Maximum tokens to generate."),
  },

  async execute(args, context) {
    // Resolve the data-forge Python package directory.
    // In a vibe-stack activated project, tools/ is linked from
    // domains/ai/data-forge/tools/, so the data-forge/ subdirectory
    // lives at .opencode/tools/data-forge/ relative to worktree.
    const dataForgeDir = path.join(context.worktree, ".opencode", "tools", "data-forge")

    // ── Build CLI argument array ───────────────────────────────────
    const subArgs: string[] = []

    const flag = (f: string, v: unknown) => {
      if (v !== undefined && v !== null) subArgs.push(f, String(v))
    }

    switch (args.subcommand) {
      case "chat": {
        flag("--prompt", args.prompt)
        flag("--provider", args.provider)
        flag("--model", args.model)
        flag("--system", args.system)
        flag("--temperature", args.temperature)
        flag("--max-tokens", args.maxTokens)
        flag("--keyfile", args.keyfile)
        break
      }
      case "describe": {
        flag("--image", args.image)
        flag("--provider", args.provider)
        flag("--model", args.model)
        flag("--prompt", args.prompt)
        flag("--temperature", args.temperature)
        flag("--max-tokens", args.maxTokens)
        flag("--keyfile", args.keyfile)
        break
      }
      case "batch-chat": {
        flag("--input-dir", args.inputDir)
        flag("--output-dir", args.outputDir)
        flag("--provider", args.provider)
        flag("--model", args.model)
        flag("--system", args.system)
        flag("--temperature", args.temperature)
        flag("--max-tokens", args.maxTokens)
        flag("--keyfile", args.keyfile)
        break
      }
      case "batch-describe": {
        flag("--input-dir", args.inputDir)
        flag("--output-dir", args.outputDir)
        flag("--provider", args.provider)
        flag("--model", args.model)
        flag("--prompt", args.prompt)
        flag("--temperature", args.temperature)
        flag("--max-tokens", args.maxTokens)
        flag("--keyfile", args.keyfile)
        break
      }
      case "list-providers": {
        flag("--keyfile", args.keyfile)
        break
      }
    }

    const cliArgs: string[] = [args.subcommand, ...subArgs]

    try {
      const result =
        await Bun.$`uv run python -m data_forge.llm ${cliArgs}`
          .cwd(dataForgeDir)
          .quiet()
          .text()

      return result.trim()
    } catch (error: unknown) {
      // Bun.ShellError carries stderr as .stderr (Buffer) — extract it
      const stderr =
        error && typeof error === "object" && "stderr" in error
          ? String((error as { stderr: Buffer }).stderr).trim()
          : ""
      return `DataForge LLM error: ${stderr || String(error)}`
    }
  },
})

export default data_forge_llm

File: tools/data-forge/src/data_forge/tools/llm.py
"""Multi-provider LLM client tool.

Reads API keys from a JSON keyfile, sends text and image prompts to
cloud LLM providers, and returns generated responses.

Supported providers: any OpenAI-compatible API (OpenAI, DeepSeek, Groq,
Together, vLLM, Ollama, etc.).

Requires:
    - ``openai`` SDK (included in core dependencies)
    - A JSON keyfile with provider credentials

Keyfile format::

    {
      "openai": {
        "api_key": "sk-...",
        "base_url": "https://api.openai.com/v1"
      },
      "deepseek": {
        "api_key": "sk-...",
        "base_url": "https://api.deepseek.com"
      }
    }
"""

from __future__ import annotations

import base64
import json
from pathlib import Path
from typing import Any

from openai import OpenAI


class LLMClient:
    """Unified client for multiple LLM providers via OpenAI-compatible APIs.

    Args:
        keyfile: Path to a JSON file mapping provider names to
            ``{api_key, base_url}`` dicts.

    Example:
        >>> client = LLMClient("api_keys.json")
        >>> client.list_providers()
        ['openai', 'deepseek']
        >>> client.chat("What is Python?", provider="openai", model="gpt-4o")
        'Python is a high-level programming language...'
    """

    def __init__(self, keyfile: str | Path) -> None:
        self._providers: dict[str, dict[str, str]] = self._load_keyfile(keyfile)
        self._clients: dict[str, OpenAI] = {}

    # ------------------------------------------------------------------
    # Keyfile management
    # ------------------------------------------------------------------

    @staticmethod
    def _load_keyfile(path: str | Path) -> dict[str, dict[str, str]]:
        """Load and validate the keyfile."""
        src = Path(path)
        if not src.is_file():
            raise FileNotFoundError(f"Keyfile not found: {src}")

        data = json.loads(src.read_text(encoding="utf-8"))

        if not isinstance(data, dict):
            raise ValueError("Keyfile must be a JSON object")

        validated: dict[str, dict[str, str]] = {}
        for name, cfg in data.items():
            if not isinstance(cfg, dict):
                raise ValueError(f"Provider '{name}' must be an object with 'api_key'")
            if "api_key" not in cfg:
                raise ValueError(
                    f"Provider '{name}' is missing required field 'api_key'"
                )
            validated[name] = {
                "api_key": cfg["api_key"],
                "base_url": cfg.get("base_url", "https://api.openai.com/v1"),
            }
        return validated

    # ------------------------------------------------------------------
    # Provider management
    # ------------------------------------------------------------------

    def list_providers(self) -> list[str]:
        """Return available provider names from the keyfile."""
        return list(self._providers.keys())

    def _get_client(self, provider: str) -> OpenAI:
        """Get or create an OpenAI client for *provider*."""
        if provider not in self._providers:
            raise ValueError(
                f"Unknown provider '{provider}'. Available: {self.list_providers()}"
            )
        if provider not in self._clients:
            cfg = self._providers[provider]
            self._clients[provider] = OpenAI(
                api_key=cfg["api_key"],
                base_url=cfg["base_url"],
            )
        return self._clients[provider]

    # ------------------------------------------------------------------
    # Text chat
    # ------------------------------------------------------------------

    def chat(
        self,
        prompt: str,
        *,
        provider: str,
        model: str,
        system: str | None = None,
        temperature: float = 0.7,
        max_tokens: int | None = None,
        **extra: Any,
    ) -> str:
        """Send a text prompt and return the model's response.

        Args:
            prompt: User message text.
            provider: Provider name (must exist in keyfile).
            model: Model name (e.g. ``"gpt-4o"``, ``"deepseek-chat"``).
            system: Optional system message.
            temperature: Sampling temperature (0-2).
            max_tokens: Maximum tokens to generate.
            **extra: Additional parameters passed to the API
                (e.g. ``top_p``, ``frequency_penalty``).

        Returns:
            The assistant's text response.
        """
        messages: list[dict[str, Any]] = []
        if system:
            messages.append({"role": "system", "content": system})
        messages.append({"role": "user", "content": prompt})

        client = self._get_client(provider)
        params: dict[str, Any] = {
            "model": model,
            "messages": messages,
            "temperature": temperature,
        }
        if max_tokens is not None:
            params["max_tokens"] = max_tokens
        params.update(extra)

        resp = client.chat.completions.create(**params)
        return resp.choices[0].message.content or ""

    # ------------------------------------------------------------------
    # Multimodal (image + text)
    # ------------------------------------------------------------------

    @staticmethod
    def _encode_image(image_path: str | Path) -> str:
        """Base64-encode an image as a data URL."""
        src = Path(image_path)
        ext = src.suffix.lower().lstrip(".")
        mime_map = {
            "jpg": "jpeg",
            "jpeg": "jpeg",
            "png": "png",
            "gif": "gif",
            "webp": "webp",
            "bmp": "bmp",
        }
        mime = mime_map.get(ext, "jpeg")
        data = base64.b64encode(src.read_bytes()).decode("utf-8")
        return f"data:image/{mime};base64,{data}"

    def describe_image(
        self,
        image_path: str | Path,
        prompt: str = "Describe this image in detail.",
        *,
        provider: str,
        model: str,
        system: str | None = None,
        temperature: float = 0.7,
        max_tokens: int | None = None,
        **extra: Any,
    ) -> str:
        """Ask a vision-capable model to describe an image.

        Args:
            image_path: Path to the image file.
            prompt: Question or instruction about the image.
            provider: Provider name.
            model: Vision-capable model name (e.g. ``"gpt-4o"``).
            system: Optional system message.
            temperature: Sampling temperature.
            max_tokens: Maximum tokens.
            **extra: Extra API parameters.

        Returns:
            The model's description text.
        """
        data_url = self._encode_image(image_path)

        content: list[dict[str, Any]] = [
            {"type": "text", "text": prompt},
            {"type": "image_url", "image_url": {"url": data_url}},
        ]

        messages: list[dict[str, Any]] = []
        if system:
            messages.append({"role": "system", "content": system})
        messages.append({"role": "user", "content": content})

        client = self._get_client(provider)
        params: dict[str, Any] = {
            "model": model,
            "messages": messages,
            "temperature": temperature,
        }
        if max_tokens is not None:
            params["max_tokens"] = max_tokens
        params.update(extra)

        resp = client.chat.completions.create(**params)
        return resp.choices[0].message.content or ""


# ======================================================================
# Convenience functions
# ======================================================================


def chat(
    prompt: str,
    *,
    keyfile: str | Path,
    provider: str,
    model: str,
    system: str | None = None,
    temperature: float = 0.7,
    max_tokens: int | None = None,
    **extra: Any,
) -> str:
    """One-shot text generation.

    Example:
        >>> chat("Hello", keyfile="keys.json", provider="openai", model="gpt-4o")
        'Hello! How can I help you?'
    """
    client = LLMClient(keyfile)
    return client.chat(
        prompt,
        provider=provider,
        model=model,
        system=system,
        temperature=temperature,
        max_tokens=max_tokens,
        **extra,
    )


def describe_image(
    image_path: str | Path,
    prompt: str = "Describe this image in detail.",
    *,
    keyfile: str | Path,
    provider: str,
    model: str,
    system: str | None = None,
    temperature: float = 0.7,
    max_tokens: int | None = None,
    **extra: Any,
) -> str:
    """One-shot image description.

    Example:
        >>> describe_image("photo.jpg", keyfile="keys.json", provider="openai", model="gpt-4o")
        'This image shows a sunset over the ocean...'
    """
    client = LLMClient(keyfile)
    return client.describe_image(
        image_path,
        prompt=prompt,
        provider=provider,
        model=model,
        system=system,
        temperature=temperature,
        max_tokens=max_tokens,
        **extra,
    )


def batch_chat(
    prompts: list[str],
    output_dir: str | Path,
    *,
    keyfile: str | Path,
    provider: str,
    model: str,
    system: str | None = None,
    temperature: float = 0.7,
    max_tokens: int | None = None,
    **extra: Any,
) -> list[Path]:
    """Batch text generation, saving each response to a file.

    Results are written to ``output_dir/result_N.txt``.

    Returns:
        Paths to generated output files.
    """
    client = LLMClient(keyfile)
    out = Path(output_dir)
    out.mkdir(parents=True, exist_ok=True)
    saved: list[Path] = []

    for i, prompt in enumerate(prompts):
        dst = out / f"result_{i}.txt"
        try:
            text = client.chat(
                prompt,
                provider=provider,
                model=model,
                system=system,
                temperature=temperature,
                max_tokens=max_tokens,
                **extra,
            )
            dst.write_text(text, encoding="utf-8")
            saved.append(dst)
        except Exception:
            continue

    return saved


def batch_describe_images(
    input_dir: str | Path,
    output_dir: str | Path,
    *,
    keyfile: str | Path,
    provider: str,
    model: str,
    prompt: str = "Describe this image in detail.",
    system: str | None = None,
    temperature: float = 0.7,
    max_tokens: int | None = None,
    **extra: Any,
) -> list[Path]:
    """Describe all images in a directory, saving each to ``<stem>.txt``.

    Returns:
        Paths to generated text files.
    """
    client = LLMClient(keyfile)
    in_dir = Path(input_dir)
    out = Path(output_dir)
    out.mkdir(parents=True, exist_ok=True)

    _supported = frozenset(
        {
            ".avif",
            ".bmp",
            ".gif",
            ".ico",
            ".jfif",
            ".jpeg",
            ".jpg",
            ".pjp",
            ".pjpeg",
            ".png",
            ".tif",
            ".tiff",
            ".webp",
        }
    )

    images = sorted(
        f for f in in_dir.iterdir() if f.is_file() and f.suffix.lower() in _supported
    )
    saved: list[Path] = []

    for img_path in images:
        dst = out / f"{img_path.stem}.txt"
        if dst.exists():
            saved.append(dst)
            continue
        try:
            text = client.describe_image(
                img_path,
                prompt=prompt,
                provider=provider,
                model=model,
                system=system,
                temperature=temperature,
                max_tokens=max_tokens,
                **extra,
            )
            dst.write_text(text, encoding="utf-8")
            saved.append(dst)
        except Exception:
            continue

    return saved
